### ISSUE TITLE
SEGFAULT for laser particles in current deposition

### ISSUE THREAD

**Issue by MaxThevenet** [ECP-WarpX/WarpX#939]
Running this [input file](https://github.com/ECP-WarpX/WarpX/files/4513679/inputs.txt) (laser propagating in vacuum, with moving window, in the lab frame) gives a reproducible SEGFAULT at iteration 458. It produces this [Backtrace](https://github.com/ECP-WarpX/WarpX/files/4513681/Backtrace.1.0.txt) when run in DEBUG mode, in current deposition, with direct and Esirkepov deposition, for order 1 or 3. I may have missed something stupid somewhere.
This happens in serial
```sh
./main2d.gnu.TPROF.MPI.OMP.ex inputs
```
or parallel jobs. Increasing the transverse resolution by a factor of 2, the bug seems to disappear.

**Comment by ax3l** [comment]
First sight: `int* ion_lev = nullptr;` lacks the `AMREX_RESTRICT` qualifier in one line before the call of the crashing `DepositCurrent`.

**Comment by ax3l** [comment]
I am not sure this implicit cast is super clean in `doEsirkepovDepositionShapeN`
```C++
    // Whether ion_lev is a null pointer (do_ionization=0) or a real pointer
    // (do_ionization=1)
    bool const do_ionization = ion_lev;
```
would do
```C++
    bool const do_ionization = (ion_lev != nullptr);
```

**Comment by ax3l** [comment]
Can reproduce the crash on Linux! :tada:

**Comment by ax3l** [comment]
The crash is in `ShapeFactor.H`:
```C++
template <>
AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
int compute_shifted_shape_factor <3> (amrex::Real* const sx,
                                      const amrex::Real x_old,
                                      const int i_new){

    using namespace amrex;

    const auto i = static_cast<int>(x_old);
    const int i_shift = i - (i_new + 1); // <-- here
```

**Comment by ax3l** [comment]
Unspecific to OpenMP :)

**Comment by LDAmorim** [comment]
The weighting is now correct.
Regardless of if we use `q_tot` or not, the run crashes on multiple nodes

**Comment by ax3l** [comment]
I think we are probably pushing `m=0` particles as if they had a mass, which generates NaNs at some point.

**Comment by ax3l** [comment]
The crash (SIGBUS) occurs here:

https://github.com/ECP-WarpX/WarpX/blob/99b8bf5ee4ac6f1212d2d98c964ca733dee52a74/Source/Particles/ShapeFactors.H#L168

`x_old` already gets into this function with a negative, very large value. `i_shift` then becomes a very large, positive value, `xint` an NaN and the array access is out-of-bounds, which causes the crash.

The particle in question has a -NaN `uzp[ip]`, `z_old`, and `vy`.

**Comment by ax3l** [comment]
The problem is in `update_laser_particle`, where we manage to flip above 1.0 for relativistic betas ([`v_over_c = 1.0075`](https://www.sciencemag.org/news/2012/06/once-again-physicists-debunk-faster-light-neutrinos)):

https://github.com/ECP-WarpX/WarpX/blob/99b8bf5ee4ac6f1212d2d98c964ca733dee52a74/Source/Laser/LaserParticleContainer.cpp#L676

Staying in expressions of momentum instead of velocity for the antenna (see: `mobility`) could be more stable.

**Comment by NeilZaim** [comment]
This was fixed by #1315 so I'm closing the issue.
